app/controllers/bio_eddie_assets_controller.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+class BioEddieAssetsController < ApplicationController
+  include ActiveStorage::SetCurrent
+
+  before_action :load_vars, except: :create
+  before_action :load_create_vars, only: :create
+
+  before_action :check_read_permission
+  before_action :check_edit_permission, only: %i(update create start_editing)
+
+  def create
+    result = BioEddieService.create_molecule(bio_eddie_params, current_user, current_team)
+
+    if result[:asset] && bio_eddie_params[:object_type] == 'Step'
+      render json: {
+        html: render_to_string(partial: 'assets/asset.html.erb', locals: {
+                                 asset: result[:asset],
+                                 gallery_view_id: bio_eddie_params[:object_id]
+                               })
+      }
+    elsif result[:asset] && bio_eddie_params[:object_type] == 'Result'
+      @my_module = result[:object].my_module
+      render json: {
+        html: render_to_string(
+          partial: 'my_modules/result.html.erb',
+            locals: { result: result[:object] }
+        )
+      }, status: :ok
+    else
+      render json: result[:asset].errors, status: :unprocessable_entity
+    end
+  end
+
+  def update
+    asset = BioEddieService.update_molecule(bio_eddie_params, current_user, current_team)
+
+    if asset
+      render json: { url: rails_representation_url(asset.medium_preview),
+                     id: asset.id,
+                     file_name: asset.blob.metadata['name'] }
+    else
+      render json: { error: t('marvinjs.no_sketches_found') }, status: :unprocessable_entity
+    end
+  end
+
+  def start_editing
+    # Activity here
+  end
+
+  private
+
+  def load_vars
+    @asset = current_team.assets.find_by(id: params[:id])
+    return render_404 unless @asset
+
+    @assoc ||= @asset.step
+    @assoc ||= @asset.result
+
+    if @assoc.instance_of?(Step)
+      @protocol = @assoc.protocol
+    elsif @assoc.instance_of?(Result)
+      @my_module = @assoc.my_module
+    end
+  end
+
+  def load_create_vars
+    @assoc = Step.find_by(id: bio_eddie_params[:object_id]) if bio_eddie_params[:object_type] == 'Step'
+    @assoc = MyModule.find_by(id: bio_eddie_params[:object_id]) if bio_eddie_params[:object_type] == 'Result'
+
+    if @assoc.instance_of?(Step)
+      @protocol = @assoc.protocol
+    elsif @assoc.instance_of?(MyModule)
+      @my_module = @assoc
+    end
+  end
```

Comment:
this could be one if block (or case statement), based on what we have in `bio_eddie_params[:object_type]`, so we don't need to check the instance_of? thing.

```
def load_create_vars
  case bio_eddie_params[:object_type]
  when 'Step'
    @assoc = Step.find_by(id: bio_eddie_params[:object_id])
    @protocol = @assoc.protocol
  when 'Result'
    @assoc = MyModule.find_by(id: bio_eddie_params[:object_id])
    @my_module = @assoc
  end
end
```